src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs

```diff
@@ -33,12 +34,13 @@ internal class TaskHost :
 #if FEATURE_APPDOMAIN
         MarshalByRefObject,
 #endif
+        IBuildEngineInternal,
         IBuildEngine9
     {
         /// <summary>
         /// True if the "secret" environment variable MSBUILDNOINPROCNODE is set.
         /// </summary>
-        private static bool s_onlyUseOutOfProcNodes = Environment.GetEnvironmentVariable("MSBUILDNOINPROCNODE") == "1";
+        private static bool s_onlyUseOutOfProcNodes = Traits.Instance.ForceOutOfProcBuildExecution;
```

Comment:
As with many escape hatches, we may get into an interesting state if different processes run with different environment variables. I don't think this change is making much of a difference, though, as we're still reading the value from the current process, potentially at a different time. To break something, there would have to be an entity setting the variable on the current running process which is very unlikely.

The ultimate fix would be to take a snapshot of the relevant environment variables in the main node, and either make them part of the handshake (connect only to nodes running with the same variables), or be sending them with requests (to avoid reading them in remote nodes altogether). Honestly, it looks like an overkill.

Reply:
Found the bug I was thinking of:
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1197992
I'm not sure we should support this sort of thing, but felt I should call it out.

Reply:
Oh, wow, if we [backported a fix for this](https://github.com/dotnet/msbuild/pull/5764) in the past, let's definitely not change the behavior. I'll revert the change. Thank you!